src/LoadBalance/LB_GatherTree.cpp

```diff
@@ -767,3 +767,275 @@ LB_GlobalPatch* LB_GatherTree( LB_PatchCount& pc, int root ) {
+   NPatch  = PatchCount.NPatchAllLv;
+}
+
+LB_GlobalTree::~LB_GlobalTree()
+{
+   delete [] Patches;
+}
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  LB_GlobalTree::IsInsidePatch
+// Description :  Check whether cell [X, Y, Z] is in patch indexed by GID
+//
+// Parameter   :  X   : global x coordinates
+//             :  Y   : global y coordinates
+//             :  Z   : global z coordinate
+//             : GID  : global ID of patch
+//
+// Return      :  true if cell [X, Y, Z] is in patch GID, false otherwise
+//-------------------------------------------------------------------------------------------------------
+bool LB_GlobalTree::IsInsidePatch(int X, int Y, int Z, long GID)  const
+{
+// sanity check
+#  ifdef GAMER_DEBUG
+   if (GID == -1)
+   {
+      Aux_Error(ERROR_INFO, "GID  == -1!\n");
+   }
+   if (GID >= NPatch)
+   {
+      Aux_Error(ERROR_INFO, "GID (%ld) >= NPatch (%ld)!\n", GID, NPatch);
+   }
+#  endif
+
+   bool IsInside = true;
+
+   int Coordinates[3] = {X, Y, Z};
+
+   for ( int l = 0; l < 3; ++l ) {
+      if (Coordinates[l] < Patches[GID].corner[l] || Coordinates[l] >=  Patches[GID].corner[l] + PS1 * amr->scale[Patches[GID].level])
+         IsInside = false;
+   }
+
+   return IsInside;
+} // LB_GlobalTree::IsInsidePatch
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  LB_GlobalTree::Local2Global
+// Description :  Convert local patch coordinates in direction XYZ to global coordinates
+//
+// Note        :  This function also works for I > PS1 and will return the global coordinate relative to GID.
+//                For patch groups, pass the GID of the root patch and I between (0 and PS2)
+//                int X = Local2Global(I, 0, GID0);
+//                int Y = Local2Global(I, 1, GID0);
+//                int Z = Local2Global(I, 2, GID0);
+//
+// Parameter   :  I   : Coordinate relative to root patch I
+//             : XYZ  : direction
+//             : GID  : Root patch relative to I
+//
+// Return      :  Global coordinate corresponding to I
+//-------------------------------------------------------------------------------------------------------
+int LB_GlobalTree::Local2Global(int I, int XYZ, long GID) const
+{
+// sanity check
+#  ifdef GAMER_DEBUG
+   if (GID == -1)
+   {
+      Aux_Error(ERROR_INFO, "GID  == -1!\n");
+   }
+   if (GID >= NPatch)
+   {
+      Aux_Error(ERROR_INFO, "GID (%ld) >= NPatch (%ld)!\n", GID, NPatch);
+   }
+#  endif
+
+   return Patches[GID].corner[XYZ] + I * amr->scale[Patches[GID].level];
+} // LB_GlobalTree::Local2Global
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  LB_GlobalTree::FindRefinedCounterpart
+// Description :  Return GID of refined patch with maximum level MaxLv (default ) that global coordinates [X, Y, Z] belong to
```

Comment:
- Remove the extra space in "(default )
- What does "default" mean here? It seems that one must specify `MaxLv`, right?

Reply:
I pass MaxLv = NLEVEL as default parameter in GatherTree.h so that by default the entire AMR tree is searched.